Punycode Arabic diacritics when not preceded by an Arabic code point in a URL host

achristensen07 opened the pull request:
#### 85e478aae63c944cb1825d6b279c2c7e189d469c
<pre>
Punycode Arabic diacritics when not preceded by an Arabic code point in a URL host
<a href="https://bugs.webkit.org/show_bug.cgi?id=247461">https://bugs.webkit.org/show_bug.cgi?id=247461</a>
rdar://100426863

Reviewed by Said Abou-Hallawa.

* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::isArabicDiacritic):
(WTF::URLHelpers::isArabicCodePoint):
(WTF::URLHelpers::isLookalikeCharacter):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/URLExtras.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/256332@main">https://commits.webkit.org/256332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2c2fdd2833df36af80a5de7e37bfe15755c55a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105038 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99443 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4750 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33458 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87826 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101120 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82069 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30541 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73380 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86523 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81823 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28126 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4378 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/40922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/84501 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/42911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39375 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19089 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WTF/wtf/URLHelpers.cpp:
```diff
@@ -191,7 +203,10 @@ static bool isLookalikeCharacter(const std::optional<UChar32>& previousCodePoint
-    
+
+    if (isArabicDiacritic(codePoint) && !isArabicCodePoint(previousCodePoint))
+        return true;
```
Can't this check be moved to a new specialized template function`isLookalikeSequence<USCRIPT_ARABIC>` and call it in the default section of the following switch statement? In this case, `isArabicDiacritic()` and `isArabicCodePoint()` can be moved also as lambda functions inside this new specialized template function.

achristensen07 replied:
The other ones are a bit different.  They check the code point before and after to make sure they are both the same kind of script.  This only needs to check the code point before.

achristensen07 replied:
Specifically, if the other code points checked by isLookalikeSequence appear at the beginning of the sequence, it's fine.  If the Arabic diacritics appear at the beginning of the sequence, it is not fine.

achristensen07 replied:
Oh, this can actually be done.  It's just moving the code to look more like the surrounding code without changing its behavior.

Commit:
Punycode Arabic diacritics when not preceded by an Arabic code point in a URL host
https://bugs.webkit.org/show_bug.cgi?id=247461
rdar://100426863

Reviewed by Said Abou-Hallawa.

* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::isArabicDiacritic):
(WTF::URLHelpers::isArabicCodePoint):
(WTF::URLHelpers::isLookalikeCharacter):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/URLExtras.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/256332@main
